### ISSUE TITLE
`sink` to remove itself from the `logger` when destructed

### ISSUE THREAD

**Issue by arsdever** [gabime/spdlog#2580]
I'm writing a simple [`QT` widget](https://github.com/arsdever/qspdlog) which should work with `spdlog` to gather and display the log messages in a `Qt` application in a pretty format. For that purpose, I've written a [custom sink](https://github.com/arsdever/qspdlog/blob/main/src/qt_logger_sink.hpp), which forwards the messages to a data model, and in the end, they are shown on the view.

The issue with my implementation is that the user needs to provide the logger to the widget for registration instead of the widget providing the user a sink to register as they want. If I try to do it in a way that the user gets the sink and registers to loggers, the sink will be kept alive (as it's a `shared_ptr`) and the underlying data (in the custom-implemented sink) will be held with the sink. Although it's possible to somehow invalidate the sink and make it do nothing, I'd like to know whether it's possible to make the sinks unregister themselves from the logger(s) they are attached to and if it was possible, why it isn't done in such a way?

In other words, I want to be able to release a sink safely.

**Comment by tt4g** [comment]
It sounds like you can achieve your goal by simply wrapping the sink you want to use with `spdlog::sinks::dist_sink`.

You can remove a sink with `remove_sink()` or replace all sinks with `set_sinks()`:

https://github.com/gabime/spdlog/blob/edc51df1bdad8667b628999394a1e7c4dc6f3658/include/spdlog/sinks/dist_sink.h#L40-L50

After creating the `std::shared_ptr<spdlog::sinks::dist_sink>` to be passed to the logger, you must copy it to a variable, and the variable must remain after passing it to the logger.

Example:

```cpp
auto dist_sink = std::make_shared<spdlog::sinks::dist_sink_mt>();
auto logger = std::make_shared<spdlog::logger>("logger", dist_sink);


std::vector<spdlog::sink_ptr> sinks;
// Push sink.
sinks.push_back(...);
// Replace sink.
dist_sink->set_sinks(std::move(sinks));
```

**Comment by arsdever** [comment]
It kinda has the solution I had in mind when I said

> Although it's possible to somehow invalidate the sink

My idea was to have the underlying data as a pointer and reset the pointer when the sink gets invalidated. As for this case, it's almost identical to clearing the vector of the sinks.

But my actual question doesn't get answered. Why wasn't it planned to provide a possibility to release the actual sink (for this case the `dist_sink` is still alive)?

**Comment by tt4g** [comment]
When `dist_sink` is no longer needed, just delete it with `std::shared_ptr::reset()`.
However, I don't understand why you want to remove `dist_sink`.
Since `spdlog::logger` does not provide the ability to safely replace sinks, why would you want to delete that `dist_sink` when you can safely replace sinks with `dist_sink`?

If you really want to delete all sink pointers referenced by `spdlog::logger`, you can do so from `spdlog::logger::sinks()`.
However, spdlog does not guarantee safety since this function directly accesses the logger's internal variables. If you delete sink while the logger is accessing this variable, it will crash.
Wiki: https://github.com/gabime/spdlog/wiki/4.-Sinks#adding-sinks-to-the-logger-after-creation

**Comment by tt4g** [comment]
I will also answer why sink cannot be safely removed from a logger. Simply because sinks are shared by multiple loggers.

**Comment by arsdever** [comment]
> When `dist_sink` is no longer needed, just delete it with `std::shared_ptr::reset()`.

Just `std::shared_ptr::reset` won't work, since when you add a sink to a logger, the logger just copies the shared pointer of the sink, increasing the ref-count of it and making it impossible to delete, unless the logger itself wants to.

> Simply because sinks are shared by multiple loggers.

So, what's wrong with that? Simply block all of the accesses to the sink, flush the content and delete.

I apologize if this shouldn't have been an issue reported to the repository. It may feel like a discussion, but I'm curious why this decision was made.

**Comment by tt4g** [comment]
> So, what's wrong with that? Simply block all of the accesses to the sink, flush the content and delete.

When you say "want to release" do you mean you want to remove sink from the logger? Or do you mean you want to delete the sink pointer?
Since the first comment said release, I assumed it meant "release the logger's sink and let the logger use the new sink".
But you seem adamant about deleting the sink pointer.

Your goal is to create a Qt widget that switches between sinks based on user interaction, right? I assume that the Qt widget creates a `dist_sink` variable in the constructor and just sets it in the logger, what's the problem?
If you delete the sink in the logger in the destructor, the reference to the `dist_sink` is gone, so there is no problem.

> I apologize if this shouldn't have been an issue reported to the repository. It may feel like a discussion, but I'm curious why this decision was made.

As far as I know, only @gabime knows the answer to this question.

**Comment by gabime** [comment]
@arsdever I am not sure I understand the question. Sinks are just normal shared ptrs. If you remove a sink from a logger, it won’t be called by the logger. If its ref count drops to zero it would be deleted. Its all up to the user. Sink is a normal shared ptr.

**Comment by arsdever** [comment]
I'll put it in the following way.

I have a sink, I register it to a couple of loggers. Then, at one point, I want to delete that sink and no longer get messages from loggers. I do have shared pointer of that sink.

1. So, one way is to traverse the logger registry, remove any existing pointer of the sink from the `logger::sinks()`.

```c++
auto& registry = spdlog::details::registry::instance();
registry.apply_all(
  [this, mysink](std::shared_ptr<spdlog::logger> logger){
    logger->sinks().erase(mysink);
  }
);
```

2. Another way is to add the sink into the `dist_sink`, and add the `dist_sink` into the loggers. With that, when I remove my sink from the `dist_sink`, it will be removed from the loggers and no longer accept messages. But the `dist_sink` will still remain in the loggers it was added to.

Why isn't there a `sink::destroy` function, which would remove the sink from everywhere?

P.S.

> If its ref count drops to zero it would be deleted.

As I understand, the only way of achieving this is by removing the pointer from the `logger::sinks()`.

**Comment by tt4g** [comment]
> Why isn't there a sink::destroy function, which would remove the sink from everywhere?

This is probably because no one needed that feature.
And there are `dist_sink` and `spdlog::logger::sinks()` as alternative methods.

I don't understand why you are concerned about the `dist_sink` remaining, why not just call `logger->sinks().erase(dist_sink)` in the destructor of the Qt widget?

**Comment by arsdever** [comment]
Sorry guys for bothering you. I feel like I got the answers to my questions. Thanks a lot for helping me in finding them.
